File: Tools/flashtool/DockWindows/UpdateWindow.py
from UpdateWindow_ui import *
from Connection import connection
from BootProtocol import BootProtocol
from msg_structs import *
from Dispatcher import dispatcher

from PyQt4.QtCore import *
from PyQt4.QtGui import *

import traceback
from Queue import Queue


# TODO:
# * Auto Update


class UpdateWindow(QMainWindow, Ui_UpdateWindow):
    KEY_FILENAME = "update_filename"

    def __init__(self):
        QMainWindow.__init__(self)
        self.setupUi(self)

        fn = QSettings().value(self.KEY_FILENAME).toString()
        self.lineEdit.setText(fn)

        model = QFileSystemModel()
        model.setRootPath(QDir.currentPath())
        model.setNameFilters(["*.hex", ])
        model.setNameFilterDisables(True)

        completer = QCompleter()
        completer.setModel(model)
        completer.setCaseSensitivity(Qt.CaseInsensitive)

        # Note: This call may be slow on some windows machines.
        # Check if there are any dead network drives or CD-ROMs
        # that take a long time to load.
        #
        self.lineEdit.setCompleter(completer)

        self.lineEdit.textChanged.connect(self.line_edit_text_changed)
        self.browseButton.clicked.connect(self.browse_button_clicked)
        self.updateButton.clicked.connect(self.update_button_clicked)

        connection.ConnectionChanged.connect(self.connection_changed)

        self.queue = Queue()
        dispatcher.subscribe(self.queue, msg_shell_to_pc.MSG_ID)

    def dragEnterEvent(self, e):
        if (len(e.mimeData().urls()) == 1 and
                e.mimeData().urls()[0].isLocalFile()):
            e.accept()

    def dropEvent(self, e):
        self.lineEdit.setText(
            e.mimeData().urls()[0].toLocalFile()
        )

    def connection_changed(self, connected):
        self.updateButton.setEnabled(connected)

    def line_edit_text_changed(self, text):
        QSettings().setValue(self.KEY_FILENAME, text)

    def browse_button_clicked(self):
        fn = str(QFileDialog.getOpenFileName(
            self, "Open file", self.lineEdit.text(),
            "Intel HEX (*.hex);;All Files (*.*)"
        ))

        if fn == "":
            return

        self.lineEdit.setText(fn)

    def update_button_clicked(self):
        # TODO: qt-event for connection change, disable update button
        #
        if dispatcher.conn is None:
            return

        class CancelError(Exception):
            pass

        def progress(percent, text):
            pd.setValue(percent)
            pd.setLabelText(text)
            QApplication.processEvents()
            if pd.wasCanceled():
                raise CancelError()

        pd = QProgressDialog(self)
        pd.setWindowModality(Qt.WindowModal)
        pd.setWindowTitle("Updating...")

        fn = str(self.lineEdit.text())
        proto = BootProtocol()

        try:
            proto.write_hex_file(fn, progress)
            QApplication.beep()
        except CancelError:
            pass
        except:
            pd.close()
            QMessageBox.critical(self, "Firmware Update Failed", traceback.format_exc(), "&OK")

File: Tools/quadcontrol/DockWindows/UpdateWindow.cpp
/**
 * Copyright (C)2015 Thomas Kindler <mail_drquad@t-kindler.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "UpdateWindow.h"
#include "ui_UpdateWindow.h"

#include <QSettings>
#include <QFileSystemModel>
#include <QCompleter>
#include <QMimeData>
#include <QFileDialog>
#include <QDebug>
#include <QProgressDialog>
#include <QMessageBox>
#include <QByteArray>
#include <QTime>

#include "MainWindow.h"
#include "IntelHexFile.h"

UpdateWindow::UpdateWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::UpdateWindow)
{
    ui->setupUi(this);

    ui->lineEdit->setText( QSettings().value("update_filename").toString() );

    connect(ui->lineEdit, &QLineEdit::textChanged, this, &UpdateWindow::lineEdit_textChanged);
    connect(ui->browseButton, &QPushButton::clicked, this, &UpdateWindow::browseButton_clicked);
    connect(ui->updateButton, &QPushButton::clicked, this, &UpdateWindow::updateButton_clicked);
    connect(&mainWindow->connection, &Connection::connectionChanged, this, &UpdateWindow::connectionChanged);

    auto model = new QFileSystemModel(this);
    model->setRootPath(QDir::currentPath());
    model->setNameFilters(QStringList { "*.hex" });
    model->setNameFilterDisables(true);

    auto completer = new QCompleter(this);
    completer->setModel(model);
    completer->setCaseSensitivity(Qt::CaseInsensitive);

    // Note: This call may be slow on some windows machines.
    // Check if there are any dead network drives or CD-ROMs
    // that take a long time to load.
    //
    ui->lineEdit->setCompleter(completer);
}

UpdateWindow::~UpdateWindow()
{
    delete ui;
}

void UpdateWindow::dragEnterEvent(QDragEnterEvent *event)
{
    if (event->mimeData()->urls().size() == 1 &&
        event->mimeData()->urls()[0].isLocalFile())
        event->acceptProposedAction();
}

void UpdateWindow::dropEvent(QDropEvent *event)
{
    ui->lineEdit->setText(
        event->mimeData()->urls()[0].toLocalFile()
    );
}

void UpdateWindow::browseButton_clicked()
{
    auto fn = QFileDialog().getOpenFileName(
        this, "Open file", ui->lineEdit->text(),
        "Intel HEX (*.hex);;All Files (*.*)"
    );

    if (fn == "")
        return;

    ui->lineEdit->setText(fn);
}

void UpdateWindow::updateButton_clicked()
{
    for (;;) {
        BootProtocol bp(mainWindow->connection, mainWindow);

        if (bp.sendHexFile(ui->lineEdit->text()))
            return;

        if (QMessageBox::critical(
                mainWindow, "Error",
                QString("Firmware update failed\n%1\n%2")
                    .arg(ui->lineEdit->text())
                    .arg(bp.errorString()),
                QMessageBox::Abort | QMessageBox::Retry
        ) != QMessageBox::Retry) {
            return;
        }
    }
}

void UpdateWindow::lineEdit_textChanged()
{
    QSettings().setValue("update_filename", ui->lineEdit->text());
}

void UpdateWindow::connectionChanged()
{
    auto c = (Connection*)sender();
    ui->updateButton->setEnabled( c->isOpen() );
}
